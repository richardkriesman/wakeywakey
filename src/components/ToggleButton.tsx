/**
 * @module components
 */

import React, { ReactNode } from "react";
import { Animated, StyleSheet, TouchableWithoutFeedback, View } from "react-native";
import Colors from "../constants/Colors";

export interface ToggleButtonProps {
    backgroundColor?: string;
    isToggled?: boolean;
    onToggle?: (isToggled: boolean) => void;
    textColorIn?: string;
    textColorOut?: string;
    title: string;
}

interface ToggleButtonState {
    isToggled: boolean;
}

const ANIMATION_TIME_MS = 75;

export class ToggleButton extends React.Component<ToggleButtonProps, ToggleButtonState> {

    public get isToggled(): boolean {
        return this.state.isToggled;
    }

    private readonly backgroundColor: string;
    private readonly colorAnimation: Animated.Value;
    private readonly textColorIn: string;
    private readonly textColorOut: string;

    public constructor(props: ToggleButtonProps) {
        super(props);
        this.backgroundColor = props.backgroundColor || Colors.appleButtonBlue;
        this.textColorIn = props.textColorIn || Colors.white;
        this.textColorOut = props.textColorOut || Colors.black;

        this.state = {
            isToggled: this.props.isToggled || false
        };
        this.colorAnimation = this.state.isToggled ? new Animated.Value(1) : new Animated.Value(0);
    }

    public render(): ReactNode {

        // build dynamic styles based on current state
        const borderDynamicStyle = {
            backgroundColor: this.backgroundColor
        };
        const containerDynamicStyle = {
            backgroundColor: this.colorAnimation.interpolate({
                inputRange: [0, 1],
                outputRange: [Colors.white, this.backgroundColor]
            })
        };
        const textDynamicStyle = {
            color: this.colorAnimation.interpolate({
                inputRange: [0, 1],
                outputRange: [this.textColorOut, this.textColorIn]
            })
        };

        // build view
        return (
            <View style={[borderDynamicStyle, styles.border]}>
                <TouchableWithoutFeedback onPress={this.onPress.bind(this)}>
                    <Animated.View style={[containerDynamicStyle, styles.container]}>
                        <Animated.Text style={textDynamicStyle}>{this.props.title}</Animated.Text>
                    </Animated.View>
                </TouchableWithoutFeedback>
            </View>
        );
    }

    private animateIn(): void {
        this.colorAnimation.setValue(0);
        Animated.timing(this.colorAnimation, {
            duration: ANIMATION_TIME_MS,
            toValue: 1
        }).start();
    }

    private animateOut(): void {
        this.colorAnimation.setValue(1);
        Animated.timing(this.colorAnimation, {
            duration: ANIMATION_TIME_MS,
            toValue: 0
        }).start();
    }

    private onPress(): void {

        // start animation
        if (!this.state.isToggled) {
            this.animateIn();
        } else {
            this.animateOut();
        }

        // update state
        this.setState({ isToggled: !this.state.isToggled }, () => { // fire listener after update
            if (this.props.onToggle) {
                this.props.onToggle(this.state.isToggled);
            }
        });
    }

}

const styles = StyleSheet.create({
    border: {
        alignItems: "center",
        borderRadius: 24,
        height: 48,
        justifyContent: "center",
        width: 48
    },
    container: {
        alignItems: "center",
        borderRadius: 22,
        height: 44,
        justifyContent: "center",
        width: 44
    }
});
