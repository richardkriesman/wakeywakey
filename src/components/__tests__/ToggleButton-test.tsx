import React from "react";
import "react-native";
import renderer from "react-test-renderer";
import {ToggleButton} from "../ToggleButton";

it("renders title correctly", () => {
    const tree = renderer.create(<ToggleButton title="OwO"/>).toJSON();
    expect(tree).toMatchSnapshot();
});

it("sets toggle state", () => {
  const tree = renderer.create(<ToggleButton title="OwO" isToggled={true} />).toJSON();
  expect(tree).toMatchSnapshot();
});
