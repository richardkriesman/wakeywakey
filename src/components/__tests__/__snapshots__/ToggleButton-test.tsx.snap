// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should toggle off 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#007aff",
      },
      Object {
        "alignItems": "center",
        "borderRadius": 24,
        "height": 48,
        "justifyContent": "center",
        "width": 48,
      },
    ]
  }
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 255, 255, 1)",
        "borderRadius": 22,
        "height": 44,
        "justifyContent": "center",
        "width": 44,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgba(0, 0, 0, 1)",
        }
      }
    >
      OwO
    </Text>
  </View>
</View>
`;

exports[`actions should toggle on 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#007aff",
      },
      Object {
        "alignItems": "center",
        "borderRadius": 24,
        "height": 48,
        "justifyContent": "center",
        "width": 48,
      },
    ]
  }
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 122, 255, 1)",
        "borderRadius": 22,
        "height": 44,
        "justifyContent": "center",
        "width": 44,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgba(255, 255, 255, 1)",
        }
      }
    >
      OwO
    </Text>
  </View>
</View>
`;

exports[`rendering initially off 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#007aff",
      },
      Object {
        "alignItems": "center",
        "borderRadius": 24,
        "height": 48,
        "justifyContent": "center",
        "width": 48,
      },
    ]
  }
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 255, 255, 1)",
        "borderRadius": 22,
        "height": 44,
        "justifyContent": "center",
        "width": 44,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgba(0, 0, 0, 1)",
        }
      }
    >
      OwO
    </Text>
  </View>
</View>
`;

exports[`rendering initially on 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#007aff",
      },
      Object {
        "alignItems": "center",
        "borderRadius": 24,
        "height": 48,
        "justifyContent": "center",
        "width": 48,
      },
    ]
  }
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 122, 255, 1)",
        "borderRadius": 22,
        "height": 44,
        "justifyContent": "center",
        "width": 44,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgba(255, 255, 255, 1)",
        }
      }
    >
      OwO
    </Text>
  </View>
</View>
`;

exports[`rendering should render the title 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#007aff",
      },
      Object {
        "alignItems": "center",
        "borderRadius": 24,
        "height": 48,
        "justifyContent": "center",
        "width": 48,
      },
    ]
  }
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 255, 255, 1)",
        "borderRadius": 22,
        "height": 44,
        "justifyContent": "center",
        "width": 44,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgba(0, 0, 0, 1)",
        }
      }
    >
      OwO
    </Text>
  </View>
</View>
`;
